multinet/uploaders/d3_json.py

```diff
@@ -0,0 +1,95 @@
+"""Multinet uploader for nested JSON files."""
+import json
+from io import StringIO
+from collections import OrderedDict
+
+from .. import db, util
+from ..errors import ValidationFailed
+from ..util import decode_data
+
+from flask import Blueprint, request
+from flask import current_app as app
+
+# Import types
+from typing import Any, List
+
+bp = Blueprint("d3_json", __name__)
+bp.before_request(util.require_db)
+
+
+def validate_d3_json(data: dict) -> List[dict]:
+    """Perform any necessary d3 json validation, and return appropriate errors."""
+    data_errors = []
+
+    # Check the structure of the uploaded file is what we expect
+    if "nodes" not in data.keys() or "links" not in data.keys():
+        data_errors.append({"error": "structure"})
+
+    # Check that links are in source -> target form
+    if not all(
+        "source" in row.keys() and "target" in row.keys() for row in data["links"]
+    ):
+        data_errors.append({"error": "invalid_link_keys"})
+
+    # Check that the keys for each dictionary match
+    if not all(data["links"][0].keys() == row.keys() for row in data["links"]):
+        data_errors.append({"error": "inconsistent_link_keys"})
+
+    # Check for duplicated nodes
+    ids = [row["id"] for row in data["nodes"]]
+    if len(ids) != len(set(ids)):
```

Comment:
Instead of `len(ids)`, you should be able to use `len(data["nodes"])`. Then, line 39 can read `ids = set(row["id"] for row in data["nodes"])` (not the lack of square brackets around the list comprehension expression).

I think this will yield the same results, but use less memory at runtime.

Reply:
Fixed in 097c5a6.